File: HackerRank/Contests/Math-Ad-Infinitum-13/movementInTreeII_moderate.hs
-- -------------------------------------------------------------------------------------
--         Author: Sourabh S Joshi (cbrghostrider); Copyright - All rights reserved.
--                       For email, run on linux (perl v5.8.5):
--   perl -e 'print pack "H*","736f75726162682e732e6a6f73686940676d61696c2e636f6d0a"'
-- -------------------------------------------------------------------------------------

-- Note that the Haskell version is slow, because of use of Integer (some tests TLE)
-- The cpp version has AC

c_MODNUM :: Integer
c_MODNUM = 1000000007

computeUnr :: Bool -> Integer -> Integer -> Integer -> Integer
computeUnr bl d k h = 
  let n     = (k - h - 1) `div` 2
      dmult = if bl == True then d else 1
      f     = d * d
      term  = dmult * (((f ^ (n+1)) - 1) `div` (f - 1))
  in  (term `mod` c_MODNUM)

computeUnrOdd  = computeUnr False
computeUnrEven = computeUnr True

unreacheable' :: Integer -> Integer -> Integer -> Integer
unreacheable' d k h 
  | ((k - h) `mod` 2) == 0 = computeUnrEven d k h
  | otherwise              = computeUnrOdd d k h

unreacheable :: Integer -> Integer -> Integer -> Integer
unreacheable d k h = if h > k then 0 else unreacheable' d k h

computeAns :: Integer -> Integer -> Integer -> Integer
computeAns d k h = 
  let dt'   = ((d ^ (k+1))-1) `div` (d-1)
      dt    = dt' `mod` c_MODNUM
      delta = unreacheable d k h 
  in  (dt - delta + c_MODNUM) `mod` c_MODNUM

runTests :: Int -> IO ()
runTests 0 = return ()
runTests tc = do
  dkhstr <- getLine
  let [d, k, h] = map read . words $ dkhstr
  let ans = computeAns d k h
  putStrLn $ show ans
  runTests (tc-1)

main :: IO ()
main = do
  tcstr <- getLine
  let tc = read tcstr
  runTests tc

File: HackerRank/Contests/Math-Ad-Infinitum-13/movementInTreeII_moderate.cpp
// -------------------------------------------------------------------------------------
//         Author: Sourabh S Joshi (cbrghostrider); Copyright - All rights reserved.
//                       For email, run on linux (perl v5.8.5):
//   perl -e 'print pack "H*","736f75726162682e732e6a6f73686940676d61696c2e636f6d0a"'
// -------------------------------------------------------------------------------------
#include <iostream>
#include <cstdio>

typedef long long Val;
const Val MODNUM = 1000000007;

Val myPow(Val x, Val n) {
    if (n == 0) return 1;
    if (n == 1) return x;
    Val subPow = myPow(x, n/2); //divide and conquer
    Val retval = subPow * subPow  ;
    retval %= MODNUM;
    retval *= (n % 2 == 1 ? x : 1);
    retval %= MODNUM;
    return retval;
}

Val findSeries(Val d, Val k) {
    if (d == 1) return ((k+1) % MODNUM); 
    Val term1 = myPow(d, k+1) + MODNUM - 1;
    term1 %= MODNUM;
    Val term2 = myPow(d - 1, MODNUM - 2);
    Val ans = term1 * term2;
    ans %= MODNUM;
    return ans;
}

Val computeUnr(bool iseven, Val d, Val k, Val h) {
    Val n = (k - h - 1) / 2;
    Val dmult = ((iseven) ? d : 1);
    Val f  = d * d; 
    f %= MODNUM;
    Val ser = findSeries(f, n);
    Val term = dmult * ser;
    term %= MODNUM;
    return term;
}

Val computeUnrEven(Val d, Val k, Val h) {
    return computeUnr(true, d, k, h);
}

Val computeUnrOdd(Val d, Val k, Val h) {
    return computeUnr(false, d, k, h);
}

Val unreacheable(Val d, Val k, Val h) {
    if (h >= k) {
        return 0;
    }
    if (((k - h) % 2) == 0) {
        return computeUnrEven(d, k, h);
    }
    return computeUnrOdd(d, k, h);
}

Val computeAns(Val d, Val k, Val h) {
    Val dt = findSeries(d, k);
    Val delta = unreacheable(d, k, h);
    Val ans = dt - delta + MODNUM;
    ans %= MODNUM;
    return ans;
}

int main() {
    Val tc; std::cin >> tc;
    for (Val t=0; t<tc; t++) {
        Val d, k, h; std::cin >> d >> k >> h;
        Val ans = computeAns(d, k, h);
        std::cout << ans << std::endl;
    }
}

